lib/Mojo/mysql/PubSub.pm

```diff
@@ -0,0 +1,291 @@
+  $payload //= '';
+  my $pid = $self->_db->pid;
+  warn "notify channel:$channel $payload listening:$pid\n" if DEBUG;
+  $self->mysql->db->query(
+    'insert into mojo_pubsub_notify(channel, payload) values (?, ?)', $channel, $payload);
+  return $self;
+}
+
+sub unlisten {
+  my ($self, $channel, $cb) = @_;
+  my $pid = $self->_db->pid;
+  warn "unlisten channel:$channel listening:$pid\n" if DEBUG;
+  my $chan = $self->{chans}{$channel};
+  @$chan = grep { $cb ne $_ } @$chan;
+  return $self if @$chan;
+  $self->mysql->db->query(
+    'delete from mojo_pubsub_subscribe where pid = ? and channel = ?', $pid, $channel);
+  delete $self->{chans}{$channel};
+  return $self;
+}
+
+sub _notifications {
+  my $self = shift;
+  my $result = $self->{db}->query(
+    'select id, channel, payload from mojo_pubsub_notify where id > ?', $self->{last_id});
+
+  while (my $row = $result->array) {
+    my ($id, $channel, $payload) = @$row;
+    $self->{last_id} = $id;
+    next unless exists $self->{chans}{$channel};
+    warn "received $id on $channel: $payload\n" if DEBUG;
+    for my $cb (@{$self->{chans}{$channel}}) { $self->$cb($payload) }
+  }
+}
+
+sub _db {
+  my $self = shift;
+
+  # Fork-safety
+  if (($self->{pid} //= $$) ne $$) {
+    my $pid = $self->{db}->pid if $self->{db};
+    warn '_DB forked pid:' . ($pid || 'N/A') . "\n" if DEBUG;
+    $self->{db}->disconnect if $pid;
+    delete @$self{qw(chans pid db)};
+  }
+
+  if ($self->{db}) {
+    warn '_DB pid:' . $self->{db}->pid, "\n" if DEBUG;
+    return $self->{db};
+  }
+
+  $self->mysql->migrations->from_data(__PACKAGE__, 'pubsub')->migrate;
+
+  my $db = $self->{db} = $self->mysql->db;
+  warn '_DB pid:' . $self->{db}->pid, "\n" if DEBUG;
+
+  if (defined $self->{last_id}) {
+    # read unread notifications
+    $self->_notifications;
+  }
+  else {
+    # get id of the last message
+    my $array = $db->query(
+      'select id from mojo_pubsub_notify order by id desc limit 1')->array;
+    $self->{last_id} = defined $array ? $array->[0] : 0;
+  }
+
+  # cleanup old subscriptions and notifications
+  $db->query(
+    'delete from mojo_pubsub_notify where ts < date_add(current_timestamp, interval -10 minute)');
+  $db->query(
+    'delete from mojo_pubsub_subscribe where ts < date_add(current_timestamp, interval -1 hour)');
+
+  # re-subscribe
+  $db->query(
+    'replace mojo_pubsub_subscribe(pid, channel) values (?, ?)', $db->pid, $_)
+    for keys %{$self->{chans}};
+
+  weaken $db->{mysql};
+  weaken $self;
+
+  my $cb;
+  $cb = sub {
+    my ($db, $err, $result) = @_;
+    if ($err) {
+      warn "wake up error: $err" if DEBUG;
+      eval { $db->disconnect };
+      delete $self->{db};
+      eval { $self->_db };
+    }
+    elsif ($self and $self->{db}) {
+      $self->_notifications;
+      $db->query('update mojo_pubsub_subscribe set ts = current_timestamp where pid = ?', $db->pid);
+      $db->query('select sleep(600)', $cb);
+    }
+  };
+  $db->query('select sleep(600)', $cb);
```

Comment:
Does this mean that both the subscriber and publisher are running the ```sleep(600)``` query? I think I would do ```$self->_db($bool)``` to indicate if this is a subscriber or not or simply refactor this method. Let me know if I'm missing out on something.